### ISSUE TITLE
"Pass by value" vs. "pass by reference" inconsistent in `fix_end_of_track()`

### ISSUE THREAD

**Issue by page200** [mido/mido#530]
`fix_end_of_track()` yields `msg` or `msg.copy()` depending on whether the previous message was a premature `'end_of_track'` message.

If I understand the difference between `msg` and `msg.copy()` correctly, the user of `fix_end_of_track()`, when trying to modify messages (for example changing the note from A to A♯) yielded by `fix_end_of_track()`, will also modify some of the notes of the original input object of `fix_end_of_track()` but not other notes. So the changes of that original object will be inconsistent, they will depend on whether the previous message was `'end_of_track'`.

We can either replace `yield msg` by `yield msg.copy()` or remove `if accum:` and `else: yield msg`.

**Comment by rdoursenaud** [comment]
I don't understand your description very well but I think this is coming from `Message` objects being mutable. IMHO `Message` objects coming from either `Ports` (MIDI Inputs) or `Tracks` (MIDI Files) should be immutable. Only objects the user creates should be mutable.

**Comment by page200** [comment]
Yes, this is about `Message` objects being mutable. Even if we make some of them immutable, the bug will still affect mutable ones. Even if `mido` didn't create any mutable ones, users' custom code or future `mido` versions might create mutable ones for some reason and the bug would resurface if not fixed.

**Comment by rdoursenaud** [comment]
Would freezing messages when associating them to a `MIDIFile` fix this?

**Comment by page200** [comment]
Even if we make it impossible for users to pass mutable "messages" as inputs to `fix_end_of_track()` (which we shouldn't), users might still modify the code when experimenting with some algorithms or improving `mido`, and then this bug would affect them.

Consistently using `msg.copy()` instead of mixing `msg` with `msg.copy()` uses more memory though (unless we use "copy on write"). So here's a new idea:

Let's have an additional input argument to `fix_end_of_track()` that decides whether the safe (default for users) or memory-friendly algorithm should be used. Then `merge_tracks()` and `MidiFile.save()` can use the memory-friendly one, and users can use the safe one by default or whatever they choose. Would you welcome this as a pull request?

**Comment by rdoursenaud** [comment]
Since this is part of the public-ish API (not prepended by `_` but not documented either), I'd accept such a PR, even though it clearly hasn't been designed for direct user interaction, rather as a facility to ensure that we don't save non-compliant files.

**Comment by page200** [comment]
(Note to myself: `merge_tracks()` has to use the safe mode because its output can be saved.)
